File: packages/sdk-python/accx/control_plane.py
"""Personal-access-token control-plane client for projects and environments."""

import json
import re
import time
import uuid
from typing import Any
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen

from .errors import AccxError

_PROJECT_UUID_RE = re.compile(r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$", re.IGNORECASE)
_VALID_LABELS = ("development", "staging", "production")


class ControlPlaneClient:
    """Projects and environments CRUD over PAT auth + freshness-headers protocol."""

    def __init__(self, *, base_url: str, personal_access_token: str, timeout: float = 15.0) -> None:
        self._base_url = base_url.rstrip("/")
        self._pat = personal_access_token
        self._timeout = min(max(timeout, 1.0), 60.0)

    def list_projects(self) -> list[dict[str, Any]]:
        payload = self._request("list_projects", {"operation": "list"})
        return payload.get("projects") if isinstance(payload.get("projects"), list) else []

    def create_project(self, name: str, *, slug: str | None = None) -> dict[str, Any]:
        body: dict[str, Any] = {"operation": "create", "name": name}
        if slug is not None:
            body["slug"] = slug
        payload = self._request("create_project", body)
        if "project" not in payload:
            raise AccxError("ACCX create_project returned no project.", status=502)
        return payload["project"]

    def rename_project(self, project: str, name: str) -> None:
        self._request("rename_project", {"operation": "rename", "projectId": self.resolve_project_id(project), "name": name})

    def delete_project(self, project: str) -> None:
        self._request("delete_project", {"operation": "delete", "projectId": self.resolve_project_id(project)})

    def list_environments(self, project: str) -> dict[str, Any]:
        target = self.resolve_project(project)
        return {"project": target, "environments": target.get("environments", [])}

    def add_environment(self, project: str, label: str) -> None:
        self._validate_label(label)
        self._request("add_environment", {"operation": "add_environment", "projectId": self.resolve_project_id(project), "label": label})

    def remove_environment(self, project: str, label: str) -> None:
        self._validate_label(label)
        self._request("remove_environment", {"operation": "remove_environment", "projectId": self.resolve_project_id(project), "label": label})

    def list_variables(self) -> list[dict[str, Any]]:
        """All variables across the workspace (metadata only, no values)."""
        payload = self._request("list_environment_variables", {"operation": "list"})
        return payload.get("variables") if isinstance(payload.get("variables"), list) else []

    def search_variables(self, query: str = "") -> list[dict[str, Any]]:
        """Greps variables across every project by key or project name (case-insensitive substring)."""
        body: dict[str, Any] = {"operation": "list"}
        if query and query.strip():
            body["query"] = query.strip()
        payload = self._request("list_environment_variables", body)
        return payload.get("variables") if isinstance(payload.get("variables"), list) else []

    def resolve_project(self, selector: str) -> dict[str, Any]:
        projects = self.list_projects()
        for p in projects:
            if p.get("id") == selector or p.get("slug") == selector or p.get("name") == selector:
                return p
        raise AccxError(f'Project "{selector}" was not found.', status=404)

    def resolve_project_id(self, selector: str) -> str:
        if _PROJECT_UUID_RE.match(selector):
            return selector
        return self.resolve_project(selector).get("id", selector)

    # ── internals ──────────────────────────────────────────────────────────

    @staticmethod
    def _validate_label(label: str) -> None:
        if label not in _VALID_LABELS:
            raise AccxError(f'Invalid environment label "{label}". Must be one of: {", ".join(_VALID_LABELS)}', status=2)

    def _request(self, command: str, body: dict[str, Any]) -> dict[str, Any]:
        url = f"{self._base_url}/api/v1/app"
        data = json.dumps({**body, "command": command}).encode("utf-8")
        nonce = uuid.uuid4().hex
        request = Request(
            url,
            data=data,
            method="POST",
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {self._pat}",
                "Origin": self._base_url,
                "X-ACCX-Request-Timestamp": str(int(time.time() * 1000)),
                "X-ACCX-Request-Nonce": nonce,
            },
        )
        return self._perform(request)

    def _perform(self, request: Request) -> dict[str, Any]:
        try:
            with urlopen(request, timeout=self._timeout) as response:
                return json.loads(response.read().decode("utf-8"))
        except HTTPError as error:
            try:
                payload = json.loads(error.read().decode("utf-8"))
                message = payload.get("error") or f"ACCX request failed with status {error.code}."
            except Exception:
                message = f"ACCX request failed with status {error.code}."
            raise AccxError(message, status=error.code) from error
        except (URLError, TimeoutError) as error:
            raise AccxError("ACCX network request failed.", status=0) from error

File: packages/sdk-js/src/control-plane.ts
import { AccxError } from "./errors.js";
import { environmentSchema, type EnvironmentLabel } from "./contracts.js";

export type CloudProject = { id: string; name: string; slug: string; createdAt: string; environments: EnvironmentLabel[] };
export type CloudEnvVar = { key: string; projectId: string; projectName: string; environmentId: string; environment: EnvironmentLabel; updatedAt: string | null; createdBy: string };
export type ControlPlaneClientOptions = {
  baseUrl: string;
  personalAccessToken: string;
  fetch?: typeof globalThis.fetch;
  timeoutMs?: number;
};

const PROJECT_UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

/**
 * Personal-access-token control-plane client. Creates, lists, renames, and
 * deletes projects and environments over the same protocol the web console uses.
 */
export class ControlPlaneClient {
  private readonly fetcher: typeof globalThis.fetch;
  private readonly baseUrl: string;
  private readonly personalAccessToken: string;
  private readonly timeoutMs: number;

  constructor(options: ControlPlaneClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/$/, "");
    this.personalAccessToken = options.personalAccessToken;
    this.fetcher = options.fetch ?? globalThis.fetch;
    this.timeoutMs = Math.min(Math.max(options.timeoutMs ?? 15_000, 1_000), 60_000);
  }

  async listProjects(): Promise<CloudProject[]> {
    const payload = await this.request<{ projects?: unknown }>("list_projects", { operation: "list" });
    return Array.isArray(payload.projects) ? (payload.projects as CloudProject[]) : [];
  }

  async createProject(name: string, slug?: string): Promise<CloudProject> {
    const payload = await this.request<{ project?: CloudProject }>("create_project", { operation: "create", name, slug });
    if (!payload.project) throw new AccxError(502, "ACCX create_project returned no project.");
    return payload.project;
  }

  async renameProject(project: string, name: string): Promise<void> {
    const projectId = await this.resolveProjectId(project);
    await this.request<{ renamed?: boolean }>("rename_project", { operation: "rename", projectId, name });
  }

  async deleteProject(project: string): Promise<void> {
    const projectId = await this.resolveProjectId(project);
    await this.request<{ deleted?: boolean }>("delete_project", { operation: "delete", projectId });
  }

  async listEnvironments(project: string): Promise<{ project: CloudProject; environments: EnvironmentLabel[] }> {
    const target = await this.resolveProject(project);
    return { project: target, environments: target.environments };
  }

  async addEnvironment(project: string, label: EnvironmentLabel): Promise<void> {
    const projectId = await this.resolveProjectId(project);
    await this.request<{ added?: boolean }>("add_environment", { operation: "add_environment", projectId, label });
  }

  async removeEnvironment(project: string, label: EnvironmentLabel): Promise<void> {
    const projectId = await this.resolveProjectId(project);
    await this.request<{ removed?: boolean }>("remove_environment", { operation: "remove_environment", projectId, label });
  }

  /**
   * Greps environment variables across every project in the workspace.
   * A query matches variable keys or project names (case-insensitive substring).
   * Results are metadata only — values never leave ACCX.
   */
  async searchVariables(query?: string): Promise<CloudEnvVar[]> {
    const payload = await this.request<{ variables?: unknown }>("list_environment_variables", {
      operation: "list",
      ...(query && query.trim().length > 0 ? { query: query.trim() } : {}),
    });
    return Array.isArray(payload.variables) ? (payload.variables as CloudEnvVar[]) : [];
  }

  /** Resolves a project selector (UUID, slug, or name) to its full record. */
  async resolveProject(selector: string): Promise<CloudProject> {
    const projects = await this.listProjects();
    const match = projects.find(p => p.id === selector || p.slug === selector || p.name === selector);
    if (!match) throw new AccxError(404, `Project "${selector}" was not found.`);
    return match;
  }

  /** Resolves a project selector (UUID, slug, or name) to its id. */
  async resolveProjectId(selector: string): Promise<string> {
    if (PROJECT_UUID_RE.test(selector)) return selector;
    return (await this.resolveProject(selector)).id;
  }

  private async request<T>(command: string, body: Record<string, unknown>): Promise<T> {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const response = await this.fetcher(`${this.baseUrl}/api/v1/app`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${this.personalAccessToken}`,
          Origin: this.baseUrl,
          "X-ACCX-Request-Timestamp": String(Date.now()),
          "X-ACCX-Request-Nonce": crypto.randomUUID().replace(/-/g, ""),
        },
        body: JSON.stringify({ command, ...body }),
        signal: controller.signal,
      });
      const payload = await response.json().catch(() => ({}));
      if (!response.ok) {
        const message = (payload as { error?: unknown }).error;
        throw new AccxError(response.status, typeof message === "string" ? message : "ACCX control-plane request was rejected.");
      }
      return payload as T;
    } finally {
      clearTimeout(timeout);
    }
  }
}

export { environmentSchema };